--ast-dump switch

dbabokin opened the pull request:
This PR adds `--ast-dump` switch, which is similar to clang's one. May be used in the following form:
- `--ast-dump=user` or `--ast-dump` to print AST for user code.
- `--ast-dump=all` to print all the code including stdlib.

The format of AST dump is intentionally made similar to format of clang AST, so it would look familiar for those who has experience with clang.

Other changes worth to note:
- `ASTNode` (and hence `Stmt` and `Expr`) class, and `Function` got member function `void Print()`, which works in the debugger. The one that was there previously didn't work in the debugger out of the box, as didn't do flashing, so nothing was printed unless `fflush()` was invoked manually.
- Most of functionality for `Print()` functions were kept the same, in some cases it was extended. But the key focus was new format. So there are things to improve in terms of what is being printed.
- `MathLib` and `CodegenOptLevel` `enum`s were changed to be `enum class` for code style uniformity.

e3m3 commented on src/ast.cpp:
```diff
@@ -48,11 +48,90 @@
+    }
+}
+
+void Indent::Print(const char *title, const SourcePos &pos) {
+    Print(title);
+    pos.Print();
+}
+
+void Indent::PrintLn(const char *title, const SourcePos &pos) {
+    Print(title, pos);
+    printf("\n");
+}
```
Why `printf()` instead of `Print()`?

dbabokin replied:
`Indent::Print()` functions are implemented using `printf()`.

`PrintLn()` does the same as `Print()`, but prints `\n` in the on top of it.

e3m3 commented on src/decl.cpp:
```diff
@@ -307,23 +314,31 @@ void Declarator::Print(int indent) const {
         for (unsigned int i = 0; i < functionParams.size(); ++i) {
-            printf("\n%*cfunc param %d:\n", indent, ' ', i);
-            functionParams[i]->Print(indent + 4);
+            char buffer[20];
+            snprintf(buffer, 20, "func param %d", i);
```
```suggestion
            static constexpr std::size_t BUFSIZE{20};
            char buffer[BUFSIZE];
            snprintf(buffer, BUFSIZE, "func param %d", i);
```

e3m3 commented on src/ast.h:
```diff
@@ -38,10 +38,59 @@
+    `-List Node #3
+
+    If the nested node needs to be annotated, setNextLabel(string) should be called
+    before recusing to the nested node.
```
```suggestion
    before recursing to the nested node.
```

e3m3 commented on src/expr.cpp:
```diff
@@ -1315,30 +1315,47 @@ int UnaryExpr::EstimateCost() const {
+        printf("postfix '++'");
+        break;
+    case PostDec: ///< Post-decrement
+        printf("postfix '++'");
```
```suggestion
        printf("postfix '--'");
```

dbabokin replied:
Good catch! Thanks!

e3m3 commented on src/expr.cpp:
```diff
@@ -2910,13 +2932,13 @@ std::pair<llvm::Constant *, bool> BinaryExpr::GetConstant(const Type *type) cons
         return "/=";
     case AssignExpr::ModAssign:
-        return "%%=";
+        return "%=";
```
Very cool :)

aneshlya commented on src/ast.cpp:
```diff
@@ -48,11 +48,90 @@
+    pos.Print();
+}
+
+void Indent::PrintLn(const char *title, const SourcePos &pos) {
```
Seems that `void Indent::PrintLn(const char *title)` method might be also useful. I see the potential uses for this, for example `indent.PrintLn("SymbolExpr: <NULL EXPR>");` instead of `indent.Print("SymbolExpr: <NULL EXPR>\n");`

dbabokin replied:
I decided to keep it simple, as in this case it might be solved with extra `\n` in the end of the passed string parameter. In case of two parameters, it can be solved only be extra `printf("\n");` call, which is bulky.

dbabokin commented on src/func.cpp:
```diff
@@ -71,6 +71,90 @@
+    indent.pushList(args.size() + 1);
+    if (args.size() > 0) {
+        for (int i = 0; i < args.size(); i++) {
+            char buffer[15];
+            snprintf(buffer, 15, "param %d", i);
```
```suggestion
            static constexpr std::size_t BUFSIZE{15};
            char buffer[BUFSIZE];
            snprintf(buffer, BUFSIZE, "param %d", i);
```

aneshlya commented:
I think we should avoid mixing `indent.Print()` and `printf()` functions, like here:
```
indent.Print("AssertStmt", pos);
printf("Message: %s\n", message.c_str());
indent.pushSingle();
```
Without looking into `Indent` class implementation, it is not clear when to use `indent.Print` and when to use `printf`.

I suggest clearly describing the expected usage in the comments for `Indent` class and adding comments for each Indent::Print method.
Also I recommend creating an issue to implement a wrapper class which can be used for all AST print operations to avoid mixing `indent.Print` and `printf` and to hide implementation details about printing from users.

e3m3 commented:
> I think we should avoid mixing `indent.Print()` and `printf()` functions, like here:
> 
> ```
> indent.Print("AssertStmt", pos);
> printf("Message: %s\n", message.c_str());
> indent.pushSingle();
> ```
> 
> Without looking into `Indent` class implementation, it is not clear when to use `indent.Print` and when to use `printf`.
> 
> I suggest clearly describing the expected usage in the comments for `Indent` class and adding comments for each Indent::Print method. Also I recommend creating an issue to implement a wrapper class which can be used for all AST print operations to avoid mixing `indent.Print` and `printf` and to hide implementation details about printing from users.

I think this was part of my confusion [here](https://github.com/ispc/ispc/pull/2299#discussion_r851554933) as well. I think for clarity, calling `Print(...)` should be `Indent::Print(this, ...)` or `this->Print(...)` in lines where the function is called without an object

dbabokin commented:
To address the confusion of mixing `indent.Print()` and `printf()` outside of `Indent` class, I added a comment in [this commit](https://github.com/ispc/ispc/pull/2299/commits/60146142505445121d44270deeff6f64a1091572).

As for mixing `Indent::Print()` and `printf()` in `Indent` implementation, I like neither of suggested improvement and addressed that in [this commit](https://github.com/ispc/ispc/pull/2299/commits/13add768e7008013682dfd900a5701b32a4a6a9e) through comments. Hope that works.
